### PR TITLE
WIP Feature/load via snowpipe

### PR DESCRIPTION
**Pull request by avinash-ankur**
## Problem
Adds the functionality of performing data transfer via snowpipe.
Snowpipe uses a Private key authentication to perform the transfer. The key will written to a file as a part of the business logic in the cartridge and its path would be added to the config target dict.  

## Proposed changes
The path to the private key file and it password, if any, is picked up as a config parameter.

## Types of changes

What types of changes does your code introduce to PipelineWise?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

### CODE REVIEW

File: target_snowflake/db_sync.py [hunk 1]
```diff
@@ -539,6 +560,107 @@ def load_csv(self, s3_key, count, size_bytes):
...
+
+        # This means Snowflake has received file and will start loading
+        assert(resp['responseCode'] == 'SUCCESS')
+
+        # Needs to wait for a while to perform transfer, delete pipe after transfer
+        while True:
+            history_resp = ingest_manager.get_history()
+
+            if len(history_resp['files']) > 0:
+                self.logger.info('Ingest Report:\n')
```
Comment (norton120) [inline target_snowflake/db_sync.py:651]: nitt: this should be one log record.

File: target_snowflake/db_sync.py [hunk 2]
```diff
@@ -539,6 +560,107 @@ def load_csv(self, s3_key, count, size_bytes):
...
+                self.logger.info(history_resp)
```
Comment (norton120) [inline target_snowflake/db_sync.py:652]: is history_resp safely formatted for logs or is it going to be a ton of noise?

File: target_snowflake/db_sync.py [hunk 3]
```diff
@@ -539,6 +560,107 @@ def load_csv(self, s3_key, count, size_bytes):
...
+                with self.open_connection() as connection:
+                    with connection.cursor() as cur:
+                        cur.execute(drop_pipe_sql)
+
+                cur.close()
```
Comment (norton120) [inline target_snowflake/db_sync.py:657]: you don't/can't close this, the `with`context manager took care of that already.

File: target_snowflake/db_sync.py [hunk 4]
```diff
@@ -539,6 +560,107 @@ def load_csv(self, s3_key, count, size_bytes):
...
+                break
+            else:
+                # wait for 20 seconds
```
Comment (norton120) [inline target_snowflake/db_sync.py:660]: should maybe step this? and log as well that we are waiting

File: target_snowflake/load_via_snowpipe.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    Encoding, \
+    PrivateFormat, \
+    NoEncryption
+from cryptography.hazmat.backends import default_backend
+from target_snowflake.s3_upload_client import S3UploadClient
+import target_snowflake.db_sync
+from snowflake.connector.errors import ProgrammingError
+# from snowflake.connector import DictCursor
+class LoadViaSnowpipe:
+
```
Comment (norton120) [inline target_snowflake/load_via_snowpipe.py:21]: if we can leave this better than we found it - some typing and a quick docstring as to what these args are would go a long way.
wtf is a dbLink?

File: target_snowflake/load_via_snowpipe.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    def __init__(self, connection_config, dbLink):
+        self.connection_config = connection_config
+        self.logger = get_logger('target_snowflake')
+        self.dbLink = dbLink
+
+    def load_via_snowpipe(self, s3_key):
+
+        # Get list if columns with types
+        columns_with_trans = [
```
Comment (norton120) [inline target_snowflake/load_via_snowpipe.py:30]: didn't I already see this code somewhere else?

File: tests/integration/test_target_snowflake.py [hunk 7]
```diff
@@ -39,7 +39,12 @@ def setUp(self):
 
         # Drop target schema
         if self.config['default_target_schema']:
-            snowflake.query("DROP SCHEMA IF EXISTS {}".format(self.config['default_target_schema']))
+            #snowflake.query("DROP SCHEMA IF EXISTS {}".format(self.config['default_target_schema']))
```
Comment (norton120) [inline tests/integration/test_target_snowflake.py:42]: let's use a test schema that we create and destroy end-to-end here.

**Review by norton120 — APPROVED** [review]
first question is... does it run & work? 
needs mega refactoring to be maintainable, but if it runs let's get it over the hump

File: target_snowflake/db_sync.py [hunk 8]
```diff
@@ -539,6 +560,107 @@ def load_csv(self, s3_key, count, size_bytes):
...
                     json.dumps({'inserts': inserts, 'updates': updates, 'size_bytes': size_bytes})))
 
+
+    def load_via_snowpipe(self, s3_key, stream):
+
+        # Get list if columns with types
+        columns_with_trans = [
+            {
+                "name": safe_column_name(name),
+                "trans": column_trans(schema)
```
[the remark this answers is not in the document]
Comment (ankur112358) [reply target_snowflake/db_sync.py:570]: I think trans here means transformation. `column_trans` checks the type of the column name and if it is binary or an array, it sets its value "trans" of the column to 'parse_json' or 'to_binary'.

**Pull request merged**